common/beerocks/bwl/common/key_value_parser.cpp

```diff
@@ -0,0 +1,225 @@
+
+#include <bwl/key_value_parser.h>
+
+#include <bcl/beerocks_state_machine.h>
+#include <bcl/beerocks_string_utils.h>
+#include <bcl/network/network_utils.h>
+
+#include <easylogging++.h>
+#include <sstream>
+
+using namespace bwl;
+
+namespace bwl {
+constexpr char EVENT_KEYLESS_PARAM_OPCODE[] = "_opcode";
+constexpr char EVENT_KEYLESS_PARAM_MAC[]    = "_mac";
+constexpr char EVENT_KEYLESS_PARAM_IFACE[]  = "_iface";
+} // namespace bwl
+
+//////////////////////////////////////////////////////////////////////////////
+////////////////////////// Local Module Functions ////////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+/**
+ * @brief Finds first occurrence of two delimiters with string between them, on given string 'str':
+ * <'delim_near'><any_characters except delimiters><'delim_far'>
+ * 
+ * @param[in] str String to find the first occurrence of two delimiters with string between them.  
+ * @param[in] pos Postion of the first characters on 'str' to be considered in the search.
+ * @param[in] delim_near Near delimiter.
+ * @param[in] delim_far Near delimiter.
+ * @return std::string::size_type Postion of the first character that matches. If no matches is 
+ * found, returns string::npos.
+ */
+static std::string::size_type find_first_of_delimiter_pair(std::string &str,
+                                                           std::string::size_type pos,
+                                                           char delim_near, char delim_far)
+{
+    // Finds first occurrence on string 'str' of
+    // <'delim_near'><any_characters except delimiters><'delim_far'>
+    auto idx = str.find_first_of(delim_far, pos);
+    if (idx == std::string::npos) {
+        return idx;
+    }
+    idx = str.rfind(delim_near, idx);
+    return idx;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/////////////////////////////// Implementation ///////////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+void KeyValueParser::parse_line(std::stringstream &ss_in, std::list<char> delimiter_list,
+                                parsed_line_t &parsed_line)
+{
+    if (delimiter_list.empty()) {
+        return;
+    }
+
+    std::string str_storage, key;
+    bool kv = true; // '1'=key, '0'=val
+    while (std::getline(ss_in, str_storage, delimiter_list.front())) {
```

Comment:
I find this recursive parse_line implementation with a delimiter list really hard to understand, for no discernable purpose... It would make a lot more sense to me to have something like this:
```
lines = str_split(reply, "\n");
for (line : lines) {
    auto idx = argument.find('=');
    if (idx != std::string::npos) {
        // store key/value
    } else {
       // store key only
    }
}
```

And to reuse it in a multiline, just move out the body that parses a single key-value into a helper function.

It could be even better with std::istream_iterator and algorithms, to avoid the copying into vectors. But: premature optimisation.

That said, it's a big change, and the current code works, so I can accept the current code.

Reply:
`str_split` is parsing the data by copying it to a vector which does reallocations, and then the vector is copied to the return value. Very wasteful function. I don't want to use it.